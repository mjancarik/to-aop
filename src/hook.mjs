import { AOP_FILTER_FUNCTION } from './symbol.mjs';

export function createHook(name, regular, callback) {
  function hook(meta) {
    return hookFor(meta, regular, callback);
  }

  hook[AOP_FILTER_FUNCTION] = function ({ property, target, prototype }) {
    return (
      (typeof regular === 'string' && property.includes(regular)) ||
      (regular instanceof RegExp && regular.test(property)) ||
      (typeof regular === 'function' &&
        regular({ property, target, name, prototype }))
    );
  };

  return {
    [name]: hook,
  };
}

export function hookFor(meta, regular, callback) {
  if (typeof regular === 'string') {
    if (meta.property.includes(regular)) {
      return callback(meta);
    }

    return null;
  }
  if (regular instanceof RegExp) {
    if (regular.test(meta.property)) {
      return callback(meta);
    }

    return null;
  }

  if (typeof regular === 'function') {
    if (regular(meta)) {
      return callback(meta);
    }

    return null;
  }

  if (Array.isArray(regular)) {
    return regular.map(({ rule, action }) => {
      return hookFor(meta, rule, action || callback);
    });
  }

  throw new TypeError(
    `Invalid rule type ${typeof regular}. Method accept string, regexp, function and array.`,
  );
}

export function hasToRegisterHook(hooks) {
  return (pattern, props) => {
    return hooks.reduce((result, hook) => {
      return (
        result ||
        (pattern[hook] &&
          typeof pattern[hook][AOP_FILTER_FUNCTION] === 'function' &&
          pattern[hook][AOP_FILTER_FUNCTION](props))
      );
    }, false);
  };
}

export const hookName = Object.freeze({
  beforeMethod: 'beforeMethod',
  afterMethod: 'afterMethod',
  aroundMethod: 'aroundMethod',
  beforeGetter: 'beforeGetter',
  afterGetter: 'afterGetter',
  aroundGetter: 'aroundGetter',
  beforeSetter: 'beforeSetter',
  afterSetter: 'afterSetter',
  aroundSetter: 'aroundSetter',
});
