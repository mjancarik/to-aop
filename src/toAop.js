'use strict';

import { AOP_PATTERN } from './symbol';
import aopForMethods from './aopForMethods';
import aopForStatic from './aopForStatic';
import createProxy from './trap/createProxy';
import createCallTrap from './trap/createCallTrap';

export function createAspect(pattern) {
  return function applyAop(target) {
    return aop(target, pattern);
  };
}

export function aop(target, pattern, settings = { constructor: false }) {
  if (settings && settings.constructor && typeof target === 'function') {
    return aopWithConstructor(target, pattern);
  }

  return applyAop(target, pattern);
}

function aopWithConstructor(target, pattern) {
  let prototype = target.prototype;

  function AOPConstructor(...rest) {
    return createCallTrap({
      target,
      object: target,
      property: 'constructor',
      pattern,
      context: this,
    })(...rest);
  }

  AOPConstructor.prototype = prototype;

  applyAop(AOPConstructor, pattern);

  return AOPConstructor;
}

export function unAop(target) {
  if (target[AOP_PATTERN]) {
    target[AOP_PATTERN] = Object.keys(target[AOP_PATTERN]).reduce(
      (pattern, hookName) => {
        pattern[hookName] = undefined;

        return pattern;
      },
      target[AOP_PATTERN]
    );

    target[AOP_PATTERN] = undefined;
  }
}

function applyAop(target, pattern) {
  if (target[AOP_PATTERN]) {
    if (typeof target === 'function') {
      aopForStatic(target, pattern);
      aopForMethods(target, pattern);

      return;
    }

    mergePattern(target, pattern);

    return;
  }

  Reflect.defineProperty(target, AOP_PATTERN, {
    value: Object.assign({}, pattern),
    enumerable: false,
    writable: true,
  });

  if (typeof target === 'function') {
    return applyAopToClass(target);
  }

  if (typeof target === 'object') {
    return applyAopToInstance(target);
  }

  throw new TypeError(
    `aop accept only object and class. You gave type of ${typeof target}.`
  );
}

function applyAopToInstance(instance) {
  return createProxy(instance);
}

function applyAopToClass(target) {
  let pattern = target[AOP_PATTERN];

  aopForStatic(target, pattern);
  aopForMethods(target, pattern);
}

function mergePattern(target, pattern) {
  let currentTargetPattern = target[AOP_PATTERN];

  target[AOP_PATTERN] = Object.entries(pattern).reduce(
    (resultPattern, [hookName, hookValue]) => {
      if (!resultPattern[hookName]) {
        resultPattern = hookName;
      }

      if (resultPattern[hookName]) {
        if (!Array.isArray(resultPattern[hookName])) {
          resultPattern[hookName] = [resultPattern[hookName]];
        }
        if (!Array.isArray(hookValue)) {
          hookValue = [hookValue];
        }

        resultPattern[hookName] = resultPattern[hookName].concat(hookValue);
      }

      return resultPattern;
    },
    currentTargetPattern
  );
}
