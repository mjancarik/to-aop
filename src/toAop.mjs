'use strict';

import { AOP_PATTERN } from './symbol.mjs';
import aopForMethods from './aopForMethods.mjs';
import aopForStatic from './aopForStatic.mjs';
import createProxy from './trap/createProxy.mjs';
import createCallTrap from './trap/createCallTrap.mjs';

export function createAspect(pattern) {
  return function applyAop(target) {
    return aop(target, pattern);
  };
}

export function aop(target, pattern, settings = { constructor: false }) {
  if (settings && settings.constructor && typeof target === 'function') {
    return aopWithConstructor(target, pattern);
  }

  return applyAop(target, pattern);
}

function aopWithConstructor(target, pattern) {
  let prototype = target.prototype;

  function AOPConstructor(...rest) {
    return createCallTrap({
      target,
      object: target,
      property: 'constructor',
      pattern,
      context: this,
    })(...rest);
  }

  AOPConstructor.prototype = prototype;

  applyAop(AOPConstructor, pattern);

  return AOPConstructor;
}

export function unAop(target) {
  if (target[AOP_PATTERN]) {
    target[AOP_PATTERN] = Object.keys(target[AOP_PATTERN]).reduce(
      (pattern, hookName) => {
        pattern[hookName] = undefined;

        return pattern;
      },
      target[AOP_PATTERN],
    );

    target[AOP_PATTERN] = undefined;
  }
}

function applyAop(target, pattern) {
  if (target[AOP_PATTERN]) {
    if (typeof target === 'function') {
      aopForStatic(target, pattern);
      aopForMethods(target, pattern);

      return target;
    }

    mergePattern(target, pattern);

    return target;
  }

  Reflect.defineProperty(target, AOP_PATTERN, {
    value: Object.assign({}, pattern),
    enumerable: false,
    writable: true,
  });

  if (typeof target === 'function') {
    return applyAopToClass(target);
  }

  if (typeof target === 'object') {
    return applyAopToInstance(target);
  }

  throw new TypeError(
    `aop accept only object and class. You gave type of ${typeof target}.`,
  );
}

function applyAopToInstance(instance) {
  return createProxy(instance);
}

function applyAopToClass(target) {
  let pattern = target[AOP_PATTERN];

  aopForStatic(target, pattern);
  aopForMethods(target, pattern);
}

function mergePattern(target, pattern) {
  let currentTargetPattern = target[AOP_PATTERN];

  target[AOP_PATTERN] = Object.entries(pattern).reduce(
    (resultPattern, [hookName, hookValue]) => {
      if (!resultPattern[hookName]) {
        resultPattern = hookName;
      }

      if (resultPattern[hookName]) {
        if (!Array.isArray(resultPattern[hookName])) {
          resultPattern[hookName] = [resultPattern[hookName]];
        }
        if (!Array.isArray(hookValue)) {
          hookValue = [hookValue];
        }

        resultPattern[hookName] = resultPattern[hookName].concat(hookValue);
      }

      return resultPattern;
    },
    currentTargetPattern,
  );
}
