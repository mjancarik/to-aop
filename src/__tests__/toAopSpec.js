import { aop, createAspect, unAop } from '../toAop';
import { hookName } from '../hook';
import createClasses from './createClasses';
import createPattern from './createPattern';

describe('createAspect method', () => {
  let withAspect = null;
  let pattern = null;

  beforeEach(() => {
    pattern = createPattern();

    withAspect = createAspect(pattern);
  });

  describe('for class', () => {
    let beforeMethod = null;
    let afterMethod = null;

    beforeEach(() => {
      beforeMethod = jest.fn();
      afterMethod = jest.fn();

      pattern = createPattern(undefined, {
        [hookName.beforeMethod]: beforeMethod,
        [hookName.afterMethod]: afterMethod,
      });

      withAspect = createAspect(pattern);
    });

    it('should call pattern.beforeMethod and pattern.afterMethod', () => {
      const variableValue = 'variableValue';
      let { A } = createClasses();
      withAspect(A);

      let instance = new A(variableValue);
      let result = instance.method({}, 1);

      expect(beforeMethod.mock.calls.length).toEqual(1);
      expect(beforeMethod.mock.calls[0]).toMatchSnapshot();
      expect(afterMethod.mock.calls.length).toEqual(1);
      expect(afterMethod.mock.calls[0]).toMatchSnapshot();

      expect(result).toEqual(variableValue);
    });

    it('should not call pattern.beforeMethod and pattern.afterMethod after unAop method is called', () => {
      let { A } = createClasses();
      withAspect(A);
      unAop(A);

      let instance = new A();
      let result = instance.method({}, 1);

      expect(beforeMethod.mock.calls.length).toEqual(0);
      expect(beforeMethod.mock.calls[0]).toMatchSnapshot();
      expect(afterMethod.mock.calls.length).toEqual(0);
      expect(afterMethod.mock.calls[0]).toMatchSnapshot();

      expect(result).toEqual(undefined);
    });

    it('should call pattern.beforeMethod and pattern.afterMethod after aop hooks are added, removed and added again to the same class', () => {
      const variableValue = 'variableValue';
      let { A } = createClasses();
      withAspect(A);
      unAop(A);
      withAspect(A);

      let instance = new A(variableValue);
      let result = instance.method({}, 1);

      expect(beforeMethod.mock.calls.length).toEqual(1);
      expect(beforeMethod.mock.calls[0]).toMatchSnapshot();
      expect(afterMethod.mock.calls.length).toEqual(1);
      expect(afterMethod.mock.calls[0]).toMatchSnapshot();

      expect(result).toEqual(variableValue);
    });

    it('should not call pattern.beforeMethod and pattern.afterMethod after aop hooks are added, removed, added and removed again from the same class', () => {
      let { A } = createClasses();
      withAspect(A);
      unAop(A);
      withAspect(A);
      unAop(A);

      let instance = new A();
      let result = instance.method({}, 1);

      expect(beforeMethod.mock.calls.length).toEqual(0);
      expect(beforeMethod.mock.calls[0]).toMatchSnapshot();
      expect(afterMethod.mock.calls.length).toEqual(0);
      expect(afterMethod.mock.calls[0]).toMatchSnapshot();

      expect(result).toEqual(undefined);
    });

    it('call pattern.beforeMethod and pattern.afterMethod with multiple aspect', () => {
      let { E, F } = createClasses();

      withAspect(E);
      withAspect(E);
      withAspect(E);
      withAspect(F);

      const value = {
        x: () => {},
      };
      const instance = new F(value);
      const result = instance.getValue();

      expect(() => result.x()).not.toThrow();
      expect(result).toEqual(value);
      expect(beforeMethod.mock.calls.length).toEqual(1);
      expect(beforeMethod.mock.calls[0]).toMatchSnapshot();
      expect(afterMethod.mock.calls.length).toEqual(1);
      expect(afterMethod.mock.calls[0]).toMatchSnapshot();
    });

    it('should not modified returnig value for static getter in parent class', () => {
      let { E, F } = createClasses();

      withAspect(E);
      withAspect(F);

      expect(E.$dependencies).toEqual(['E']);
      expect(F.$dependencies).toEqual(['F']);
    });

    it('should not modified returnig value for static getter in child class', () => {
      let { E, F } = createClasses();

      withAspect(F);
      withAspect(E);

      expect(E.$dependencies).toEqual(['E']);
      expect(F.$dependencies).toEqual(['F']);
    });

    it('should call pattern.beforeMethod and pattern.afterMethod for settings constructor: true', () => {
      let { A } = createClasses();
      let AA = aop(A, pattern, { constructor: true });
      let instance = new AA(1);
      let result = instance.method({}, 1);

      expect(beforeMethod.mock.calls.length).toEqual(2);
      expect(beforeMethod.mock.calls[0]).toMatchSnapshot();
      expect(afterMethod.mock.calls.length).toEqual(2);
      expect(afterMethod.mock.calls[0]).toMatchSnapshot();
      expect(result).toEqual(1);
    });

    it('should call pattern.beforeMethod and pattern.afterMethod for settings constructor: true and extended class', () => {
      let { B } = createClasses();
      let BB = aop(B, pattern, { constructor: true });
      let instance = new BB(1);
      let result = instance.method({}, 1);

      expect(beforeMethod.mock.calls.length).toEqual(2);
      expect(beforeMethod.mock.calls[0]).toMatchSnapshot();
      expect(afterMethod.mock.calls.length).toEqual(2);
      expect(afterMethod.mock.calls[0]).toMatchSnapshot();
      expect(result).toEqual('B 1');
    });
  });

  describe('for instance', () => {
    let beforeMethod = null;
    let beforeGetter = null;
    let afterMethod = null;

    beforeEach(() => {
      beforeMethod = jest.fn();
      beforeGetter = jest.fn();
      afterMethod = jest.fn();

      pattern = createPattern(undefined, {
        [hookName.beforeMethod]: beforeMethod,
        [hookName.beforeGetter]: beforeGetter,
        [hookName.afterMethod]: afterMethod,
      });

      withAspect = createAspect(pattern);
    });

    it('should call pattern.beforeMethod and pattern.afterMethod', () => {
      const variableValue = 'variableValue';
      let { A } = createClasses();
      let instance = withAspect(new A(variableValue));
      let result = instance.method({}, 1);

      expect(beforeMethod.mock.calls.length).toEqual(1);
      expect(beforeMethod.mock.calls[0]).toMatchSnapshot();
      expect(afterMethod.mock.calls.length).toEqual(1);
      expect(afterMethod.mock.calls[0]).toMatchSnapshot();

      expect(result).toEqual(variableValue);
    });

    it('should not call pattern.beforeMethod and pattern.afterMethod after unAop method is called', () => {
      let { A } = createClasses();
      let a = new A();
      let instance = withAspect(a);
      unAop(a);
      let result = instance.method({}, 1);

      expect(beforeMethod.mock.calls.length).toEqual(0);
      expect(beforeMethod.mock.calls[0]).toMatchSnapshot();
      expect(afterMethod.mock.calls.length).toEqual(0);
      expect(afterMethod.mock.calls[0]).toMatchSnapshot();

      expect(result).toEqual(undefined);
    });

    it('should call pattern.beforeMethod and pattern.afterMethod after aop hooks are added, removed and added again to the same method', () => {
      const variableValue = 'variableValue';
      let { A } = createClasses();
      let a = new A(variableValue);
      withAspect(a);
      unAop(a);
      let instance = withAspect(a);
      let result = instance.method({}, 1);

      expect(beforeMethod.mock.calls.length).toEqual(1);
      expect(beforeMethod.mock.calls[0]).toMatchSnapshot();
      expect(afterMethod.mock.calls.length).toEqual(1);
      expect(afterMethod.mock.calls[0]).toMatchSnapshot();

      expect(result).toEqual(variableValue);
    });

    it('should not call pattern.beforeMethod and pattern.afterMethod after aop hooks are added, removed, added and removed again from the same method', () => {
      let { A } = createClasses();
      let a = new A();
      withAspect(a);
      unAop(a);
      let instance = withAspect(a);
      unAop(a);
      let result = instance.method({}, 1);

      expect(beforeMethod.mock.calls.length).toEqual(0);
      expect(beforeMethod.mock.calls[0]).toMatchSnapshot();
      expect(afterMethod.mock.calls.length).toEqual(0);
      expect(afterMethod.mock.calls[0]).toMatchSnapshot();

      expect(result).toEqual(undefined);
    });

    it('should call pattern.beforeGetter and pattern.beforeMethod', () => {
      let { C } = createClasses();
      let instance = withAspect(new C());
      let entries = instance.map.entries();

      expect(beforeGetter.mock.calls.length).toEqual(2);
      expect(beforeMethod.mock.calls.length).toEqual(1);
      expect(entries.next().done).toEqual(true);
    });

    it('should return value from instance getter', () => {
      let { F } = createClasses();

      const value = 1;
      let instance = withAspect(new F(value));

      expect(instance.value).toEqual(value);
      expect(beforeGetter.mock.calls.length).toEqual(1);
    });
  });

  describe('for object', () => {
    let beforeMethod = null;
    let afterMethod = null;

    beforeEach(() => {
      beforeMethod = jest.fn();
      afterMethod = jest.fn();

      pattern = createPattern(undefined, {
        [hookName.beforeMethod]: beforeMethod,
        [hookName.afterMethod]: afterMethod,
      });

      withAspect = createAspect(pattern);
    });

    it('should call pattern.beforeMethod and pattern.afterMethod for dummy object', () => {
      let dummy = {
        method() {},
      };

      let xdummy = withAspect(dummy);
      xdummy.method('works');

      expect(beforeMethod.mock.calls.length).toEqual(1);
      expect(afterMethod.mock.calls.length).toEqual(1);
    });

    it('should call pattern.beforeMethod and pattern.afterMethod for global console', () => {
      spyOn(console, 'log');

      let xconsole = withAspect(console);
      xconsole.log('works');

      expect(beforeMethod.mock.calls.length).toEqual(1);
      expect(afterMethod.mock.calls.length).toEqual(1);
    });

    it('should call pattern.beforeMethod and pattern.afterMethod for object with context', () => {
      const object = {
        value: 1,
        method: function increase() {
          this.value++;
        },
      };

      const xobject = withAspect(object);

      xobject.method();

      expect(xobject.value).toEqual(2);
      expect(beforeMethod.mock.calls.length).toEqual(1);
      expect(afterMethod.mock.calls.length).toEqual(1);
    });
  });
});
