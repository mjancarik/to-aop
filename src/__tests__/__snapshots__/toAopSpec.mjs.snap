// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createAspect method for class call pattern.beforeMethod and pattern.afterMethod with multiple aspect 1`] = `
[
  {
    "args": [],
    "context": F {
      "_value": {
        "x": [Function],
      },
    },
    "meta": {},
    "object": {
      "getValue": [Function],
    },
    "property": "getValue",
    "target": [Function],
  },
]
`;

exports[`createAspect method for class call pattern.beforeMethod and pattern.afterMethod with multiple aspect 2`] = `
[
  {
    "args": [],
    "context": F {
      "_value": {
        "x": [Function],
      },
    },
    "meta": {},
    "object": {
      "getValue": [Function],
    },
    "payload": {
      "x": [Function],
    },
    "property": "getValue",
    "target": [Function],
  },
]
`;

exports[`createAspect method for class should call pattern.beforeMethod and pattern.afterMethod 1`] = `
[
  {
    "args": [
      {},
      1,
    ],
    "context": A {
      "variable": "variableValue",
    },
    "meta": {},
    "object": {
      "_privateMethod": [Function],
      "method": [Function],
      "superMethod": [Function],
    },
    "property": "method",
    "target": [Function],
  },
]
`;

exports[`createAspect method for class should call pattern.beforeMethod and pattern.afterMethod 2`] = `
[
  {
    "args": [
      {},
      1,
    ],
    "context": A {
      "variable": "variableValue",
    },
    "meta": {},
    "object": {
      "_privateMethod": [Function],
      "method": [Function],
      "superMethod": [Function],
    },
    "payload": "variableValue",
    "property": "method",
    "target": [Function],
  },
]
`;

exports[`createAspect method for class should call pattern.beforeMethod and pattern.afterMethod after aop hooks are added, removed and added again to the same class 1`] = `
[
  {
    "args": [
      {},
      1,
    ],
    "context": A {
      "variable": "variableValue",
    },
    "meta": {},
    "object": {
      "_privateMethod": [Function],
      "method": [Function],
      "superMethod": [Function],
    },
    "property": "method",
    "target": [Function],
  },
]
`;

exports[`createAspect method for class should call pattern.beforeMethod and pattern.afterMethod after aop hooks are added, removed and added again to the same class 2`] = `
[
  {
    "args": [
      {},
      1,
    ],
    "context": A {
      "variable": "variableValue",
    },
    "meta": {},
    "object": {
      "_privateMethod": [Function],
      "method": [Function],
      "superMethod": [Function],
    },
    "payload": "variableValue",
    "property": "method",
    "target": [Function],
  },
]
`;

exports[`createAspect method for class should call pattern.beforeMethod and pattern.afterMethod for settings constructor: true 1`] = `
[
  {
    "args": [
      1,
    ],
    "context": A {},
    "meta": {},
    "object": [Function],
    "property": "constructor",
    "target": [Function],
  },
]
`;

exports[`createAspect method for class should call pattern.beforeMethod and pattern.afterMethod for settings constructor: true 2`] = `
[
  {
    "args": [
      1,
    ],
    "context": A {},
    "meta": {},
    "object": [Function],
    "payload": A {
      "variable": 1,
    },
    "property": "constructor",
    "target": [Function],
  },
]
`;

exports[`createAspect method for class should call pattern.beforeMethod and pattern.afterMethod for settings constructor: true and extended class 1`] = `
[
  {
    "args": [
      1,
    ],
    "context": B {},
    "meta": {},
    "object": [Function],
    "property": "constructor",
    "target": [Function],
  },
]
`;

exports[`createAspect method for class should call pattern.beforeMethod and pattern.afterMethod for settings constructor: true and extended class 2`] = `
[
  {
    "args": [
      1,
    ],
    "context": B {},
    "meta": {},
    "object": [Function],
    "payload": B {
      "variable": 1,
    },
    "property": "constructor",
    "target": [Function],
  },
]
`;

exports[`createAspect method for class should not call pattern.beforeMethod and pattern.afterMethod after aop hooks are added, removed, added and removed again from the same class 1`] = `undefined`;

exports[`createAspect method for class should not call pattern.beforeMethod and pattern.afterMethod after aop hooks are added, removed, added and removed again from the same class 2`] = `undefined`;

exports[`createAspect method for class should not call pattern.beforeMethod and pattern.afterMethod after unAop method is called 1`] = `undefined`;

exports[`createAspect method for class should not call pattern.beforeMethod and pattern.afterMethod after unAop method is called 2`] = `undefined`;

exports[`createAspect method for instance should call pattern.beforeMethod and pattern.afterMethod 1`] = `
[
  {
    "args": [
      {},
      1,
    ],
    "context": A {
      "variable": "variableValue",
    },
    "meta": {},
    "object": A {
      "variable": "variableValue",
    },
    "property": "method",
    "target": [Function],
  },
]
`;

exports[`createAspect method for instance should call pattern.beforeMethod and pattern.afterMethod 2`] = `
[
  {
    "args": [
      {},
      1,
    ],
    "context": A {
      "variable": "variableValue",
    },
    "meta": {},
    "object": A {
      "variable": "variableValue",
    },
    "payload": "variableValue",
    "property": "method",
    "target": [Function],
  },
]
`;

exports[`createAspect method for instance should call pattern.beforeMethod and pattern.afterMethod after aop hooks are added, removed and added again to the same method 1`] = `
[
  {
    "args": [
      {},
      1,
    ],
    "context": A {
      "variable": "variableValue",
    },
    "meta": {},
    "object": A {
      "variable": "variableValue",
    },
    "property": "method",
    "target": [Function],
  },
]
`;

exports[`createAspect method for instance should call pattern.beforeMethod and pattern.afterMethod after aop hooks are added, removed and added again to the same method 2`] = `
[
  {
    "args": [
      {},
      1,
    ],
    "context": A {
      "variable": "variableValue",
    },
    "meta": {},
    "object": A {
      "variable": "variableValue",
    },
    "payload": "variableValue",
    "property": "method",
    "target": [Function],
  },
]
`;

exports[`createAspect method for instance should not call pattern.beforeMethod and pattern.afterMethod after aop hooks are added, removed, added and removed again from the same method 1`] = `undefined`;

exports[`createAspect method for instance should not call pattern.beforeMethod and pattern.afterMethod after aop hooks are added, removed, added and removed again from the same method 2`] = `undefined`;

exports[`createAspect method for instance should not call pattern.beforeMethod and pattern.afterMethod after unAop method is called 1`] = `undefined`;

exports[`createAspect method for instance should not call pattern.beforeMethod and pattern.afterMethod after unAop method is called 2`] = `undefined`;
